### PR TITLE
add failpoint coverage for semantic task edge cases

### LINKED ISSUE #149: Add failpoint-based fault injection for high-value semantic task tests
## Enhancement

Introduce failpoint-based fault injection for the highest-value concurrency and failure-path tests, starting with semantic task lease handling and related worker/runtime paths.

### Motivation

We currently have a few tests that still depend on real time passing or on carefully orchestrated async timing windows. Those tests are serviceable, but they are harder to make deterministic and can become flaky under slow CI or small timing shifts.

If we decide to adopt failpoint in drive9, the highest-value targets are not generic sleep replacement everywhere, but the places where we need precise control over:

- lease-expiry timing
- renew vs stop races
- handler panic / retry / recover boundaries
- storage / provider failure injection on critical semantic task paths

### Highest-value initial scope

1. `pkg/datastore/semantic_tasks.go`
   - inject deterministic control over lease-related "now" decisions used by renew / ack / retry / recover paths
   - use this to replace real-time-dependent lease-expiry tests such as renew-expired behavior

2. `pkg/server/semantic_worker.go`
   - inject controlled races around renew, shutdown, lease loss, ack/retry boundaries, and panic cleanup
   - use this to validate lease-lost behavior without relying only on timing-sensitive coordination

3. `pkg/backend/image_extract.go`
   - inject load / extractor / writeback failures at precise points
   - improve coverage of long-running semantic task failure behavior and retry semantics

### Why these areas first

These paths have the best return on investment for failpoint:

- they are correctness-sensitive
- they involve asynchronous state transitions or concurrency boundaries
- they are harder to test deterministically with ordinary mocks alone
- they are more valuable than converting every small `time.Sleep(...)` in the repo

### Non-goals

- Do not convert the whole repository to failpoint at once
- Do not use failpoint as a blanket replacement for simple polling or straightforward test doubles
- Do not introduce failpoint in low-value areas before the semantic task / datastore / image extract paths are covered

### Suggested rollout

1. Add failpoint toolchain / documentation for local and CI use
2. Start with `pkg/datastore/semantic_tasks.go` lease-expiry and lease-ownership tests
3. Extend to `pkg/server/semantic_worker.go` race-oriented tests
4. Add targeted failure injection for `pkg/backend/image_extract.go`
5. Re-evaluate whether additional packages justify failpoint adoption

### Expected outcome

- fewer timing-sensitive flaky tests in high-value semantic task paths
- better deterministic coverage for lease-loss, recover, panic, and retry edge cases
- a reusable fault-injection mechanism for future long-running task types such as audio extraction

### PR DESCRIPTION
**Pull request by JaySon-Huang**
## Summary
- add failpoint runner infrastructure, local documentation, and CI coverage so failpoint tests can run consistently without overlapping normal `go test` runs
- add deterministic datastore and semantic worker failpoint tests for lease expiry, renew, finalize, lease-loss, and panic cleanup boundaries
- add backend image extract writeback failpoint tests to verify transactional rollback when content-text updates or embed-task bridging fail

Closes #149.

## Test plan
- [x] `go test -v ./pkg/server/...`
- [x] `go test -v ./pkg/backend/...`
- [x] `python3 scripts/run_failpoint_tests.py`

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **New Features**
  * Added failpoint testing infrastructure to simulate failures and validate rollback/recovery.

* **Documentation**
  * Added failpoint testing guidance to developer docs and README (how to write and run failpoint tests).

* **Tests**
  * Added comprehensive failpoint-tagged tests for image extraction and semantic task lease/processing; removed some legacy tests replaced by failpoint variants.

* **Chores**
  * Added a failpoint test runner and Make target; CI job timeout increased to accommodate the new tests; module metadata updated.

### CODE REVIEW

File: pkg/datastore/semantic_tasks.go [hunk 1]
```diff
@@ -547,6 +549,22 @@ func (s *Store) semanticTaskLeaseError(ctx context.Context, taskID string) error
...
+		switch injected := val.(type) {
+		case string:
+			parsed, err := time.Parse(time.RFC3339Nano, injected)
+			if err == nil {
+				now = parsed.UTC()
+			}
+		}
+	})
+	return now
+}
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply pkg/datastore/semantic_tasks.go:566]: Thanks for the suggestion. I agree that narrower failpoint scopes can be a good follow-up when multiple lease-sensitive operations need to be injected independently in the same package.

For this PR, I’m keeping the single `semanticTaskLeaseNow` injection point as-is for two reasons:

1. The current tests enable the failpoint only within a single test scope and always disable it via `t.Cleanup`, so there is no cross-test leakage in the current usage.
2. The three call sites are all exercising the same underlying concern: overriding the lease-comparison “now” used by ownership-sensitive task transitions. Splitting this immediately into `ack` / `retry` / `recover` variants would add API surface and duplicated helper logic without materially improving determinism for the cases covered in this PR.

So I’d prefer to keep the shared clock override in this change set and revisit finer-grained failpoint names only if we add tests that need overlapping or independently-targeted injections for those operations.

**Review by qiffang — COMMENTED** [review]
## Code Review — PR #157: add failpoint coverage for semantic task edge cases

CI is green. Overall this is a well-structured PR that replaces flaky sleep-based timing tests with deterministic failpoint injection. Good work. A few observations and concerns below.

---

### Positives

- **Deterministic lease boundaries**: Replacing `time.Sleep` + poll loops with `semanticTaskLeaseNow` failpoint injection is a clear win for test reliability. The ack/retry/recover tests in `datastore` are clean and easy to follow.
- **Good separation**: `*_failpoint_test.go` files with `//go:build failpoint` keeps the instrumented tests isolated from normal `go test` runs.
- **Infrastructure is solid**: The Python runner (`scripts/run_failpoint_tests.py`) with enable/disable in a `try/finally` block is the right pattern. CI integration is straightforward.
- **Documentation is thorough**: Both `AGENTS.md` and `README.md` updates explain the "why" and the caveats (no parallel with `go test`, no parallel with lint).

---

### Issues

#### 1. `semanticTaskStillOwned` uses `time.Now()` instead of `semanticTaskLeaseNow()` (Bug)

**`pkg/server/semantic_worker.go`** — the new `semanticTaskStillOwned` method does:
```go
now := time.Now().UTC()
```

But the ack/retry paths in `datastore` now use `semanticTaskLeaseNow()` which is failpoint-injectable. This means the ownership re-check and the actual ack/retry use **different clocks** under failpoint. In production the difference is nil, but it creates an inconsistency in the test model — the finalize ownership check could pass with real time while the underlying ack uses injected time (or vice versa). Consider using `semanticTaskLeaseNow()` here too, or accepting the injected time as a parameter, so the ownership check and the terminal mutation see the same `now`.

#### 2. `semanticTaskStillOwned` silently returns `(false, nil)` on `sql.ErrNoRows` — but `COUNT(*)` never returns `ErrNoRows`

```go
if errors.Is(err, sql.ErrNoRows) {
    return false, nil
}
```

`SELECT COUNT(*)` always returns exactly one row (the count, possibly 0). This branch is dead code. It's not harmful, but it's misleading — it suggests the author expected a row-scan failure that can't happen. Remove it or change the query to `SELECT 1 ... LIMIT 1` if you actually want the `ErrNoRows` path.

#### 3. `semanticTaskStillOwned` returns `(false, nil)` when `store == nil || task == nil` — caller treats this as "not owned"

In `processTask`, the call site:
```go
stillOwned, err := m.semanticTaskStillOwned(...)
if err != nil { /* log and continue */ }
else if !stillOwned { return }
```

If `store` or `task` is nil (shouldn't happen in practice, but the guard is there), it returns `(false, nil)` which causes the worker to silently return without acking or retrying — dropping the task on the floor. A nil guard that causes silent data loss is worse than a panic. If these are truly impossible, remove the guard; if they're possible, return an error.

#### 4. Ownership check error is treated as "still owned" — should it be?

```go
if err != nil {
    logger.Warn(...)
    // falls through to ack/retry
} else if !stillOwned {
    return
}
```

When the ownership check query fails (network blip, context cancelled), the code falls through and attempts ack/retry anyway. This is arguably correct (the ack/retry themselves have lease checks), but it means the ownership pre-check only prevents mutations on success. Document this decision — it's a reasonable one, but the intent isn't obvious.

#### 5. Helper types moved from `_test.go` to `_failpoint_test.go` break non-failpoint builds

`gatedServerImageExtractor` and the `assertNoObservedLogForTask` helper were removed from `semantic_worker_test.go` and placed in `semantic_worker_failpoint_test.go` (which has `//go:build failpoint`). If any non-failpoint test still references these types, it will break under `go test` without `-tags=failpoint`. CI passed so this seems fine, but double-check that the old tests that used `gatedServerImageExtractor` (renew, lease-lost, panic) are **all** migrated — which they appear to be. The `panicServerImageExtractor` was renamed to `gatedPanicServerImageExtractor` with added `release` channel — make sure the behavioral change (blocking on `release` before panicking) is intentional and doesn't mask issues in the panic cleanup test.

#### 6. `run_failpoint_tests.py` — shell injection in `command_exists`

```python
def command_exists(name: str) -> bool:
    return subprocess.run(
        ["bash", "-lc", f"command -v {name} >/dev/null 2>&1"],
    ...
```

The `name` parameter is interpolated into a shell string. Currently it's hardcoded to `"podman"` so there's no real risk, but if this function is ever reused with user input, it's injectable. Use `shlex.quote(name)` or pass `name` as an argument via `["bash", "-lc", "command -v \"$1\" >/dev/null 2>&1", "_", name]`.

#### 7. Minor: import ordering in `semantic_worker.go`

`"github.com/pingcap/failpoint"` is placed after the project imports, breaking the standard Go import grouping (stdlib → external → project). Run `goimports` or move it to the external group.

---

### Questions

1. The `semanticWorkerLeaseRenewStopped` failpoint is injected but I don't see any test that enables it — is it reserved for future use?
2. The `injectBeforeSemanticTaskFinalize` method exists only to call `failpoint.InjectCall`. Why not inline the `failpoint.InjectCall` directly in `processTask`? The indirection adds a method that has no non-failpoint purpose.

---

### Verdict

The test coverage improvement is valuable and the infrastructure is well-built. Fix the `time.Now()` vs `semanticTaskLeaseNow()` inconsistency in `semanticTaskStillOwned`, clean up the dead `ErrNoRows` branch, and consider the nil-guard semantics. Everything else is minor.

LGTM with the above addressed.

**Comment by JaySon-Huang** [comment]
Addressed in `71d4d87`.

Updates in this follow-up:
- aligned the worker-side ownership re-check with a failpoint-aware clock via `semanticWorkerLeaseNow()`
- removed the dead `sql.ErrNoRows` path in `semanticTaskStillOwned`
- changed the `nil` guard in `semanticTaskStillOwned` to return an explicit error instead of silently treating it as lease loss
- added an inline comment documenting why ownership-check read failures still fall back to datastore ack/retry lease validation
- tightened the failpoint test helper lookup to disambiguate by `task_type`
- removed the unused `semanticWorkerLeaseRenewStopped` failpoint hook
- documented the known `command_exists()` shell-injection tradeoff in the failpoint runner
- cleaned up import grouping in `semantic_worker.go`

Validation:
- `make build`
- `make test TEST_P=1`
- `make test-failpoint`
- `make lint`

Note: plain `make test` was flaky in this environment with package-level parallelism because the MySQL/testcontainers path hit intermittent `unexpected EOF` / `invalid connection`. Re-running with `TEST_P=1` passed cleanly, and the targeted package tests plus failpoint suite also passed.

File: scripts/run_failpoint_tests.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        / "bin"
+        / "failpoint-ctl"
+    )
+    if not failpoint_ctl.exists():
+        print(f"failpoint-ctl not found at {failpoint_ctl}", file=sys.stderr)
+        print(
+            "Install it with: go install github.com/pingcap/failpoint/failpoint-ctl@latest",
+            file=sys.stderr,
+        )
+        return 1
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply scripts/run_failpoint_tests.py]: Addressed in `9a2e0db`.

I updated the runner to resolve `failpoint-ctl` via a dedicated `resolve_failpoint_ctl_path()` helper that prefers `GOBIN` and otherwise falls back to the first `GOPATH` entry, so it matches `go install` behavior on both configured-`GOBIN` and multi-entry-`GOPATH` setups.

File: scripts/run_failpoint_tests.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            run(
+                ["go", "test", "-v", "-tags", "failpoint", pkg, "-run", pattern],
+                cwd=repo_root,
+                env=env,
+            )
+    except subprocess.CalledProcessError as exc:
+        exit_code = exc.returncode or 1
+    finally:
+        subprocess.run([str(failpoint_ctl), "disable", "."], cwd=repo_root, check=False, env=env)
+
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply scripts/run_failpoint_tests.py:68]: Addressed in `9a2e0db`.

The `finally` cleanup now captures the result of `failpoint-ctl disable .`, prints any stdout/stderr on failure, and turns the script exit status non-zero if disable fails. That way the runner can no longer report success while leaving the tree instrumented.

**Review by qiffang — APPROVED** [review]
